### PR TITLE
Float(oo) -> oo; float(oo) -> inf

### LINKED ISSUE #6965: Be smarter about Float('inf') and oo
See http://stackoverflow.com/q/16905103/161801 .  An integral fails if you use inf instead of oo. This is related to issue 6127 .

Original issue for #6965: http://code.google.com/p/sympy/issues/detail?id=3866
Original author: https://code.google.com/u/asmeurer@gmail.com/
Referenced issues: #6127

### LINKED ISSUE #8910: Comparing Float Infinity with Core Infinity
So while looking at Issue #8906 I came across a possibly related issue.

```
>>> from sympy.stats import *
>>> from sympy import Symbol
>>> x = Symbol("x")
>>> X = Uniform("u", 0, 1)
>>> P(X<.5)
Traceback (most recent call last):
  ...
    raise TypeError("cannot determine truth value of\n%s" % self)
TypeError: cannot determine truth value of
-oo < -inf
```

The Traceback involved intersecting some sets to figure out the domain of X, which is where the comparison `-oo < -inf` arose. I'll try to fix it quickly.

### LINKED ISSUE #15352: Solveset does not find solution when using Eq with stats
`solveset(Eq(sympy.stats.P(sympy.stats.Normal(Dummy(),0,1)>0.2),Symbol("x")))`
The solution should just be `sympy.stats.P(sympy.stats.Normal(sympy.Dummy(),0,1)>0.2)`
which is
`{-erf(0.1*sqrt(2))/2 + erf(inf*sqrt(2))/2}`
but the actual output is
`EmptySet()`

When the above is rewritten without Eq
`>>> solveset(sympy.stats.P(sympy.stats.Normal(Dummy(),0,1)>0.2)-Symbol("x"))`
`{-erf(0.1*sqrt(2))/2 + erf(inf*sqrt(2))/2}`
which is the correct solution.

This was tested with Sympy versions 1.3 and 1.4.dev

### LINKED ISSUE #16720: Poly(1.*x**2 - 1.*oo, x).to_exact() generates ZeroDivisionError in polys
This works
```python
>>> Poly(1.*x**2 - 1.*oo)
Poly(1.0*x**2 + 1.0*(-inf), x, -inf, domain='RR')
>>> _.to_exact()
```
But when the x is given explicitly, it doesn't:
```python
Poly(x**2 + (-inf), x, -inf, domain='QQ')
>>> Poly(1.*x**2 - 1.*oo, x)
Poly(1.0*x**2 - inf, x, domain='RR')
>>> _.to_exact()
Traceback (most recent call last):
  File "C:\Users\leslie\Desktop\Dropbox\jnk.py", line 46, in <module>
    p.to_exact()
  File "polys\polytools.py", line 739, in to_exact
    result = f.rep.to_exact()
  File "polys\polyclasses.py", line 293, in to_exact
    return f.convert(f.dom.get_exact())
  File "polys\polyclasses.py", line 300, in convert
    return DMP(dmp_convert(f.rep, f.lev, f.dom, dom), dom, f.lev)
  File "polys\densebasic.py", line 563, in dmp_convert
    return dup_convert(f, K0, K1)
  File "polys\densebasic.py", line 540, in dup_convert
    return dup_strip([ K1.convert(c, K0) for c in f ])
  File "polys\densebasic.py", line 540, in <listcomp>
    return dup_strip([ K1.convert(c, K0) for c in f ])
  File "polys\domains\domain.py", line 107, in convert
    return self.convert_from(element, base)
  File "polys\domains\domain.py", line 97, in convert_from
    result = _convert(element, base)
  File "polys\domains\pythonrationalfield.py", line 61, in from_RealField
    p, q = K0.to_rational(a)
  File "polys\domains\realfield.py", line 100, in to_rational
    return self._context.to_rational(element, limit)
  File "polys\domains\mpelements.py", line 133, in to_rational
    a = n//d
ZeroDivisionError: integer division or modulo by zero
```

### PR DESCRIPTION
**Pull request by smichr**
sympification of inf will now give oo and this
avoids issues like that in #16720

The same applies to nan.

All interactions of floating nan, inf and -inf with
sympy objects will coerce the float into a symPy
version.

#### References to other Issues or PRs

cf 
#6127 closed issue about nan and NaN being merged
#6216 closed -+inf with pretty printer and Float('inf')
closes #6965 {Be smarter about Float('inf') and oo} because any sympified input will be oo not float('inf')
#7765 closed oo*Float doesn't return oo (related to Point instantiation)
closes #7771 {Change multiplication behavior of oo with Float to return oo} as alternate
#8451 Infinity equals infinity, or not?
closes #8910 {Comparing Float Infinity with Core Infinity} there is only 1 infinity now
closes #15352 {Solveset does not find solution when using Eq with stats} there in only 1 infinity now
closes #16720 Poly(1.*x**2 - 1.*oo, x).to_exact() generates ZeroDivisionError in polys
#### Other comments

Historical note: nan was changed to NaN but the same was never done with oo.

Someone able to interpret the ast changes should take a look to make sure that I've captured the intent there.

In addition, we need to make sure that we want `float(oo) == oo`. Since we have `1.0 == 1` it seems like we do, but currently in master, `Float(oo) != oo`.

#### Release Notes

* core
    * numbers - Float(oo) now gives oo; inf can be created with float(oo), if needed

### CODE REVIEW

File: sympy/core/numbers.py [hunk 1]
```diff
@@ -954,14 +954,20 @@ def __new__(cls, num, dps=None, prec=None, precision=None):
                 num = '0' + num
             elif num.startswith('-.') and len(num) > 2:
                 num = '-0.' + num[2:]
+            elif num == 'inf' or num == '+inf':
+                return S.Infinity
+            elif num == '-inf':
+                return S.NegativeInfinity
         elif isinstance(num, float) and num == 0:
             num = '0'
+        elif isinstance(num, float) and 1/num == 0:
```
Comment (smichr) [inline sympy/core/numbers.py]: Is this an ok way to test for inf or -inf?

**Comment by smichr** [comment]
ping @jksuom 
It looks like RR and CC need to work with the mpf values. Does this look right to you?

**Comment by jksuom** [comment]
I think that RR and CC should reject infinite values and nan. There are no such elements in other domains. That could probably be done by adding `__new__` methods to `RealElement` and `ComplexElement`  that would check the values created by `_mpf.__new__` or `_mpc.__new__`.
(Maybe that could also be done by `__init__`.)

File: sympy/core/numbers.py [hunk 2]
```diff
@@ -2698,6 +2714,10 @@ def __sub__(self, other):
                 return S.Infinity
         return NotImplemented
 
+    @_sympifyit('other', NotImplemented)
+    def __rsub__(self, other):
+        return (-self).__sub__(other)
```
Comment (jksuom) [inline sympy/core/numbers.py]: `__add__`?

File: sympy/core/numbers.py [hunk 3]
```diff
@@ -2914,13 +2936,17 @@ def __sub__(self, other):
...
+                    return S.NaN
                 else:
                     return Float('-inf')
             else:
                 return S.NegativeInfinity
         return NotImplemented
 
+    @_sympifyit('other', NotImplemented)
+    def __rsub__(self, other):
+        return (-self).__sub__(other)
```
Comment (jksuom) [inline sympy/core/numbers.py]: ?

File: sympy/polys/domains/complexfield.py [hunk 4]
```diff
@@ -69,6 +69,10 @@ def from_sympy(self, expr):
         real, imag = number.as_real_imag()
 
         if real.is_Number and imag.is_Number:
+            if number is oo:
+                return inf+0.0j
```
Comment (jksuom) [inline sympy/polys/domains/complexfield.py]: I think that infinities and nan should raise an exception here, and in `RR.from_sympy` as well. CoercionFailed would be the proper choice.

Comment (jksuom) [reply sympy/polys/domains/complexfield.py]: It is probably not necessary to make the checks in `from_sympy`. It would suffice that ComplexElement should raise CoercionFailed.

File: sympy/core/numbers.py (hunk 3 above)
Comment (smichr) [reply sympy/core/numbers.py]: Yes, this should be `__add__`. The purpose is to keep `1.0 - +/-oo` from becoming `-float(+/-oo) + 1.0`.

**Comment by smichr** [comment]
> There are no such elements in other domains.
Should it be allowed in either of these?
```python
assert (oo in EX) is True
assert (oo in RR[x, y]) is True
```

**Comment by jksuom** [comment]
IMHO they should not be allowed in any domain that is defined in polys/domains.

**Comment by asmeurer** [comment]
There was some discussion on merging the float infinities with the symbolic ones, so that `oo` is what it is now, but also a Float subclass with the appropriate `_mpf_` and so on. Is this moving towards that goal? I think having a separate float inf that is rarely encountered isn't great.

**Comment by asmeurer** [comment]
Otherwise, this PR doesn't seem to make much sense. If there is a separate Float infinity, then surely the `Float` constructor should return it. What do `Float('inf')` or `Float(float('inf'))` give in this branch?

**Comment by asmeurer** [comment]
I just checked and they all give `oo`. Also in this branch `float(oo)` gives `float('inf')` (that is, the native Python type, not a SymPy type, can `float` return anything that isn't a `float` type?). So is this branch abolishing the floating point infinity entirely? I think this is good, but it might still be a good idea to make `oo` itself a Float, so that `Float('inf')._mpf_` and so on works.

File: sympy/codegen/ast.py [hunk 5]
```diff
@@ -130,12 +130,16 @@
...
 from sympy.core.compatibility import string_types
 from sympy.core.expr import Expr
-from sympy.core.numbers import Float, Integer
+from sympy.core.numbers import Float, Integer, oo
 from sympy.core.relational import Lt, Le, Ge, Gt
 from sympy.core.sympify import _sympify, sympify, SympifyError
 from sympy.utilities.iterables import iterable
 
 
+inf = float(oo)
```
Comment (asmeurer) [inline sympy/codegen/ast.py]: I don't think it's good to have these native Python floats floating around (so to speak). Especially since `oo` is a singleton and often compared with `is`, I think it could cause issues.

File: sympy/codegen/ast.py [hunk 6]
```diff
@@ -130,12 +130,16 @@
...
+noo = -oo
```
Comment (asmeurer) [inline sympy/codegen/ast.py]: `noo` is just as long as `-oo`, but harder to read.

File: sympy/codegen/ast.py [hunk 7]
```diff
@@ -130,12 +130,16 @@
...
+ninf = float(noo)
```
Comment (asmeurer) [inline sympy/codegen/ast.py]: Same with `-inf` (but again, I don't think any of these should be here).

**Comment by asmeurer** [comment]
Maybe something like

```diff
diff --git a/sympy/core/numbers.py b/sympy/core/numbers.py
index fb483a9270..d30a72024c 100644
--- a/sympy/core/numbers.py
+++ b/sympy/core/numbers.py
@@ -2629,7 +2629,7 @@ def __abs__():
         return S.Half


-class Infinity(with_metaclass(Singleton, Number)):
+class Infinity(with_metaclass(Singleton, Float)):
     r"""Positive infinite quantity.

     In real analysis the symbol `\infty` denotes an unbounded
@@ -2676,7 +2676,10 @@ class Infinity(with_metaclass(Singleton, Number)):
     __slots__ = []

     def __new__(cls):
-        return AtomicExpr.__new__(cls)
+        obj = AtomicExpr.__new__(cls)
+        obj._mpf_ = (0, 0, -456, -2)
+        obj._prec = -1
+        return obj

     def _latex(self, printer):
         return r"\infty"
```

(and ditto for NegativeInfinity). The `_mpf_` tuple is from `Float('inf')._mpf_` on master. 

I'm not 100% if that's the correct way to write `__new__` or if `AtomicExpr.__new__` is what should be used there. There also seem to be some issues with the tests with this, so some more may be required to make this work.

**Comment by smichr** [comment]
> ```diff
> class Infinity(with_metaclass(Singleton, Float)):
> ```

I really want to avoid blurring the distinction between `float(oo)` and `oo`. The only way, now, to get a floating `inf` is with `float(oo)`. And from `oo` (as from `pi`) you can get an mpf of arbitrary precision:
```
>>> oo._as_mpf_val(1)
(0, 0, -456, -2)
>>> oo._as_mpf_val(111)
(0, 0, -456, -2)
>>> pi._as_mpf_val(2)
(0, 3, 0, 2)
>>> pi._as_mpf_val(22)
(0, 1647099, -19, 21)
```
It should now be impossible to generate a floating nan, inf or ninf in a SymPy expression.

**Comment by asmeurer** [comment]
`float(oo)` is the same as `float('inf')`. It's a Python `float` object, not a SymPy object. So I don't think it should be used anywhere (it doesn't know anything about arbitrary precision, for instance). 

I'm not sure if pi is a reasonable comparison. As far as I understand, mpmath only has one floating point infinity, `(0, 0, -456, -2)`. -456 and -2 are just numbers that were chosen as fill values. See https://github.com/fredrik-johansson/mpmath/blob/e0ec6e894465c276448d42669f0b838071662199/mpmath/libmp/libmpf.py#L92-L95. pi on the other hand will give a different `_mpf_` (Float) for every precision. So it's not accurate to say that `pi` is a Float, any more than `sin(1)` or `Rational(1, 10)`. But oo has a unique floating point representation, regardless of the precision. So we might as well make it *be* that float instance. In master we have such an instance already, `Float('inf')`, but there's no good reason to have it duplicate `oo`. 

It may also be worth looking into the discussions from when `Float('nan')` was removed. Apparently we never made `S.NaN` a Float (float nan is also a little odd in that it has nonstandard equality behavior, so maybe it doesn't make sense to treat it the same as infinity). 

Maybe we can get away with not having a floating point infinity at all, and just using `oo`. But my impression is that things like the evalf code will be easier to maintain if everything that's a float, including the infinities, have meaningful _mpf_ values. I don't see any changes to evalf.py in this branch, so perhaps that impression is wrong.

**Comment by smichr** [comment]
> Is this moving towards that goal?

It is there: no sympified expression can contain `float(oo)`. The only place `float(oo)` appears is in tests and in codegen as the output value for `oo`. `oo` is SymPy, `float(oo) == float('inf')`. Writing that now, I see that a change that can be made is to internally write `float(oo)` as `float('inf')` so it doesn't have to pass through SymPy.

> perhaps that impression is wrong

Every Number needs to be able to be evalf'ed to an arbitrary precision. Nothing has changed in that regard. `oo` has, when needed, the value you indicated and mpmath knows what to do with it.

With the discussion about "real" and "extended real", the current definition of `oo` fits in nicely: reals are all Floats; extended reals are `{-oo, Floats, +oo}`.

File: sympy/core/numbers.py [hunk 8]
```diff
@@ -2798,10 +2787,10 @@ def __hash__(self):
         return super(Infinity, self).__hash__()
 
     def __eq__(self, other):
-        return other is S.Infinity
+        return other is S.Infinity or other == _inf
```
Comment (jksuom) [inline sympy/core/numbers.py]: `float('inf')` also here?

Comment (smichr) [reply sympy/core/numbers.py]: They should all be gone now.

**Comment by smichr** [comment]
Should Float reject `oo` as input? 
In master, `Float('oo').is_Float == True` but `oo.is_Float` is False. 
In master there is `Float(oo)`, `oo` and `float(oo`) -- 2 SymPy and 1 Python; 
now there is only `oo` and `float(oo)`.

Integer and Rational reject `oo` as input.

**Comment by jksuom** [comment]
> Should Float reject oo as input?

I don't think that would be necessary as Float is based on mpf that has inf defined. Excluding oo from RR is another matter.

**Comment by smichr** [comment]
> I don't think that would be necessary

I was thinking of the nice invariant that `Float(_).is_Float` would always be true. On the other hand, `Triangle(collinear points) -> Line` so the morphing of input is not exceptional.

As I was trying to say above, conceptually I like keeping `oo` out of the Floats. That's what I meant when drawing the comparison with `pi`. Neither are currently Floats; both know what their mpf value is, however, and mpmath knows what to do with them.

Is there more to do on this PR or do you think, if tests pass, that this is ready?

**Comment by jksuom** [comment]
> Float(_).is_Float would always be true.

That would be satisfied by oo if [Aaron's suggestion](https://github.com/sympy/sympy/pull/16727#issuecomment-487182859) is implemented. (I don't know if that would be helpful somewhere, but it wouldn't probably be harmful, either.) However, there is still `Float(nan)`...

> Is there more to do on this PR or do you think, if tests pass, that this is ready?

I like these changes. There may be more to do but it should not be necessary to include everything in one PR. So I think this is ready.

**Comment by smichr** [comment]
> there is still `Float(nan)`...

Yes...the little trilogy of numbers-that-aren't-numbers: nan, oo, -oo.

Conceptually, I think this takes us a step forward by eliminating one of two SymPy infinite values. So let's commit this. It has been something discussed for some time and it can be improved upon if necessary.

**Comment by asmeurer** [comment]
> As I was trying to say above, conceptually I like keeping oo out of the Floats. That's what I meant when drawing the comparison with pi. Neither are currently Floats; both know what their mpf value is, however, and mpmath knows what to do with them.

I'm not sure what the original question is. The string `'oo'` (as opposed to `'inf'`) is nonstandard for floats, so doesn't necessarily need to be supported (though I think it's fine if it is). 

As for accepting `Float(oo)` at all, unlike `pi`, +/- infinity *is* a float, according to the mpf definition (which is based on IEEE 754). We purged `float('nan')` from the codebase a long time ago in favor of the symbolic NaN, which we never made into a Float, and it hasn't caused any issues. Floating point nan also has some annoying semantics which we wanted to avoid for a symbolic system (in some lesser sense, floating point `inf` has some semantics we don't want either, for instance, we use `1/0 == zoo` instead).

In 10 years, I personally have never once seen any SymPy operation produce `Float('inf')`. So externally at least, it seems it isn't needed. I don't know if having `oo` be a true Float with an `_mpf_` would make internal code cleaner, such as the code in evalf.py.

In short, perhaps we can leave this as is for now, until a need for `oo` to be a Float becomes clear. It never happened with NaN, and maybe it will never happen with `oo` either.
